### PR TITLE
PG16 compatibility - varnullingrels additions

### PR DESCRIPTION
**Pull request by naisila**
PG16 introduced a new entry varnnullingrels to Var, which represents our partkey in pg_dist_partition.
This commit does the necessary changes for this.

Relevant PG commit:
https://github.com/postgres/postgres/commit/2489d76c4906f4461a364ca8ad7e0751ead8aa0d 2489d76c4906f4461a364ca8ad7e0751ead8aa0d

Checkout main PG16 PR for CI tests: particularly the upgrade tests are successful. #6952

### CODE REVIEW

File: src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/12.1-1.sql [hunk 1]
```diff
@@ -106,6 +106,19 @@ BEGIN
...
+
+    -- if we are upgrading from PG14/PG15 to PG16+,
+    -- we need to add varnullingrels to partkey if it's not already there
+    IF substring(current_Setting('server_version'), '\d+')::int >= 16 THEN
+    EXECUTE $cmd$
+        UPDATE pg_catalog.pg_dist_partition
+        SET partkey = (split_part(partkey, 'varlevelsup', 1)
+                       || 'varnullingrels (b) :varlevelsup'
+                       || split_part(partkey, 'varlevelsup', 2))
+        WHERE partkey IS NOT NULL AND partkey NOT LIKE '%varnullingrels%';
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/12.1-1.sql:118]: should we write with `ILIKE` just to be sure that there are no edge cases with capital letters?

File: src/test/regress/expected/citus_local_tables.out [hunk 2]
```diff
@@ -635,7 +635,7 @@ FROM pg_dist_partition WHERE logicalrelid = 'citus_local_table_4'::regclass;
 SELECT column_name_to_column('citus_local_table_4', 'a');
                                                   column_name_to_column
 ---------------------------------------------------------------------
```
Comment (onderkalaci) [inline src/test/regress/expected/citus_local_tables.out:637]: do we have any tests with column_to_column_name in PG 16? If possible, would be nice to add to upgrade tests to show that everything is fine for all version upgrades

File: src/test/regress/expected/citus_local_tables_mx.out [hunk 3]
```diff
@@ -769,8 +769,8 @@ SELECT logicalrelid, partmethod, partkey FROM pg_dist_partition
         ORDER BY logicalrelid;
      logicalrelid     | partmethod |                                                          partkey
 ---------------------------------------------------------------------
- parent_dropped_col   | h          | {VAR :varno 1 :varattno 1 :vartype 1082 :vartypmod -1 :varcollid 0 :varlevelsup 0 :varnosyn 1 :varattnosyn 1 :location -1}
- parent_dropped_col_2 | h          | {VAR :varno 1 :varattno 5 :vartype 23 :vartypmod -1 :varcollid 0 :varlevelsup 0 :varnosyn 1 :varattnosyn 5 :location -1}
+ parent_dropped_col   | h          | {VAR :varno 1 :varattno 1 :vartype 1082 :vartypmod -1 :varcollid 0 :varnullingrels (b) :varlevelsup 0 :varnosyn 1 :varattnosyn 1 :location -1}
```
Comment (onderkalaci) [inline src/test/regress/expected/citus_local_tables_mx.out:772]: I'm confused. How are these changing with PG 15? Are we manually changing the test outputs? 

For these kinds of tests, can't we use `column_to_column_name` such that they are accurate for all versions?

File: src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/12.1-1.sql (hunk 1 above)
Comment (naisila) [reply src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/12.1-1.sql:118]: I didn't know about `ilike` and `like` difference before, good idea, thanks.

File: src/test/regress/expected/citus_local_tables.out (hunk 2 above)
Comment (naisila) [reply src/test/regress/expected/citus_local_tables.out:637]: Adding upgrade tests for this makes sense, I will do that.

File: src/test/regress/expected/citus_local_tables_mx.out (hunk 3 above)
Comment (naisila) [reply src/test/regress/expected/citus_local_tables_mx.out:772]: I need the normalization rule for the column_name_to_column tests. I may as well keep it for these tests as well.

File: src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/latest.sql [hunk 4]
```diff
@@ -106,6 +106,19 @@ BEGIN
     -- restore citus catalog tables
     --
     INSERT INTO pg_catalog.pg_dist_partition SELECT * FROM public.pg_dist_partition;
+
+    -- if we are upgrading from PG14/PG15 to PG16+,
+    -- we need to add varnullingrels to partkey if it's not already there
+    IF substring(current_Setting('server_version'), '\d+')::int >= 16 THEN
+    EXECUTE $cmd$
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/latest.sql:113]: it is slightly concerning to update the pg_dist_partition like this... I mean, I cannot see anything wrong now, but I'm also not very happy about this.

One idea that I'm considering is to fully rely on PG. Use PG to store the column name before the upgrade, and then later, use the column name to set back the column in the new PG version.

On citus_prepare_pg_upgrade, we can use `column_to_column_name` to record the column names for tables:

```
CREATE TABLE public.pg_dist_partkeys_pre_16 AS SELECT logicalrelid, column_to_column_name(partkey) as col_name FROM pg_dist_partition WHERE partkey IS NOT NULL AND partkey NOT ILIKE '%varnullingrels%';; 
```

and then here:

UPDATE pg_dist_partition SET partkey = column_name_to_column(col_name) FROM pg_dist_partkeys_pre_16 WHERE pg_dist_parkeys.logicalrelid = pg_dist_partition.logicalrelid;

what do you think about that?

Comment (naisila) [reply src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/latest.sql:113]: I see your point.
Let me try. Maybe I don't need to put extra conditions either and just apply this in upgrades.

Comment (naisila) [reply src/backend/distributed/sql/udfs/citus_finish_pg_upgrade/latest.sql]: Check out the successful tests in this commit https://github.com/citusdata/citus/pull/6952/commits/342c146100007eefd1a69c329b1830de848a8743

File: src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/12.1-1.sql [hunk 5]
```diff
@@ -84,6 +84,13 @@ BEGIN
        objects.colocationid
     FROM pg_catalog.pg_dist_object objects,
          pg_catalog.pg_identify_object_as_address(objects.classid, objects.objid, objects.objsubid) address;
+
+    -- if we are upgrading from PG14/PG15 to PG16+,
+    -- we will need to regenerate the partkeys because they will include varnullingrels as well.
+    -- so we save the partkeys as column names here
+    CREATE TABLE IF NOT EXISTS public.pg_dist_partkeys_pre_upgrade AS
```
Comment (onderkalaci) [inline src/backend/distributed/sql/udfs/citus_prepare_pg_upgrade/12.1-1.sql:91]: I'm still in favor of calling this pg_dist_partkeys_pre_upgrade_16, and checking if `varnullingrels` exists for the row.

I'm suggesting that because, say if you are upgrading from PG16 to PG17, we are needlessly modifying the rows.

is there any problem with that?

**Review by onderkalaci — APPROVED** [review]
thanks, looks good now, slightly tricky change but it seems like the only way (e.g., other than storing Var in the catalog -- we should have stored the column name)
